File: python/src/assesslite/lattice.py
"""Assumption lattice (AssessLite core spec v0.2, spec/lattice.md).

The pooling invariances (pool across clusters / time) are the "stronger symmetry ->
one number" commitments. This refits the exposure estimate under every pool-or-stratify
combination and reports whether the conclusion depends on those pooling commitments.
Kept identical to the R engine.
"""
from __future__ import annotations

from itertools import combinations

import numpy as np

from .estimator import fit_estimate


def build_lattice(assessment) -> dict:
    s = assessment.structure
    axes, var_of = [], {}
    if s["cluster"] is not None:
        axes.append("cluster_exchangeability"); var_of["cluster_exchangeability"] = s["cluster"]
    if s["time"] is not None:
        axes.append("temporal_translation"); var_of["temporal_translation"] = s["time"]

    if not axes:
        return {"axes": [], "variables": {}, "nodes": [], "verdict": "not_resolvable",
                "reading": "no poolable structural axes (cluster or time) were declared; "
                           "the pooling lattice is empty"}

    main = assessment.estimate
    k = len(axes)
    nodes = []
    for r in range(k + 1):
        for P in combinations(axes, r):
            P = list(P)
            stratified = [ax for ax in axes if ax not in P]
            fit = fit_estimate(assessment.analysis, assessment.data,
                               strata=[var_of[a] for a in stratified])
            if fit is None:
                continue
            same_sign = np.sign(fit["value"]) == np.sign(main["value"]) or np.sign(main["value"]) == 0
            excl_null = fit["ci_low"] > 0 or fit["ci_high"] < 0
            status = ("reversed" if (not same_sign and excl_null)
                      else "attenuated" if not excl_null else "consistent")
            nodes.append({"pooled": P, "stratified": stratified, "n_pooled": len(P),
                          "estimate": fit["value"], "ci_low": fit["ci_low"],
                          "ci_high": fit["ci_high"], "n": fit["n"], "status": status})

    statuses = [nd["status"] for nd in nodes]
    verdict = ("unstable" if "reversed" in statuses
               else "not_resolvable" if "attenuated" in statuses else "stable")
    words = ", ".join(a.replace("_", " ") for a in axes)
    if verdict == "stable":
        reading = (f"the exposure estimate keeps the same direction and stays distinguishable from "
                   f"no effect under every pool-or-stratify choice over {{{words}}}; the conclusion "
                   f"does not depend on these pooling commitments")
    elif verdict == "not_resolvable":
        reading = (f"under some pool-or-stratify choices over {{{words}}} the interval comes to include "
                   f"no effect: the direction holds throughout, but whether the effect is resolved "
                   f"depends on the pooling commitments")
    else:
        reading = (f"the exposure estimate changes sign under some pool-or-stratify choice over "
                   f"{{{words}}}; the conclusion depends on these pooling commitments and does not "
                   f"hold across the lattice")
    return {"axes": axes, "variables": var_of, "nodes": nodes, "verdict": verdict, "reading": reading}

File: R/lattice.R
# Assumption lattice (AssessLite core spec v0.2, spec/lattice.md).
# The pooling invariances (pool across clusters / time) are the "stronger symmetry
# -> one number" commitments. This refits the exposure estimate under every
# pool-or-stratify combination and reports whether the conclusion depends on those
# pooling commitments. Nodes are ordered by how much is pooled; the top node (pool
# everything) is the main estimate, the bottom node (stratify everything) makes the
# weakest pooling assumptions.

#' Build the pooling assumption lattice
#'
#' @param audit a structural_audit object (with a fitted estimate).
#' @return the audit with a $lattice element.
assumption_lattice <- function(audit) {
  stopifnot(inherits(audit, "structural_audit"))
  s <- audit$structure
  axis_invariance <- character()
  axis_var <- character()
  if (!is.null(s$cluster)) { axis_invariance <- c(axis_invariance, "cluster_exchangeability"); axis_var <- c(axis_var, s$cluster) }
  if (!is.null(s$time))    { axis_invariance <- c(axis_invariance, "temporal_translation");   axis_var <- c(axis_var, s$time) }
  names(axis_var) <- axis_invariance
  axes <- axis_invariance

  if (length(axes) == 0) {
    audit$lattice <- list(axes = character(), variables = list(), nodes = list(),
      verdict = "not_resolvable",
      reading = "no poolable structural axes (cluster or time) were declared; the pooling lattice is empty")
    return(audit)
  }

  main <- audit$estimate
  k <- length(axes)
  subsets <- lapply(0:(2^k - 1), function(m) axes[bitwAnd(m, 2^(seq_len(k) - 1)) > 0])
  nodes <- list()
  for (P in subsets) {
    stratified <- setdiff(axes, P)
    fit <- fit_estimate(audit, audit$data, strata = unname(axis_var[stratified]))
    if (is.null(fit)) next
    same_sign <- sign(fit$value) == sign(main$value) || sign(main$value) == 0
    excl_null <- fit$ci_low > 0 || fit$ci_high < 0
    status <- if (!same_sign && excl_null) "reversed"
              else if (!excl_null) "attenuated"
              else "consistent"
    nodes[[length(nodes) + 1]] <- list(
      pooled = I(as.character(P)), stratified = I(as.character(stratified)),
      n_pooled = length(P), estimate = fit$value,
      ci_low = fit$ci_low, ci_high = fit$ci_high, n = fit$n, status = status)
  }

  statuses <- vapply(nodes, function(x) x$status, character(1))
  verdict <- if (any(statuses == "reversed")) "unstable"
             else if (any(statuses == "attenuated")) "not_resolvable"
             else "stable"
  axis_words <- gsub("_", " ", axes)
  reading <- switch(verdict,
    stable = sprintf(paste0("the exposure estimate keeps the same direction and stays distinguishable ",
      "from no effect under every pool-or-stratify choice over {%s}; the conclusion does not depend on ",
      "these pooling commitments"), paste(axis_words, collapse = ", ")),
    not_resolvable = sprintf(paste0("under some pool-or-stratify choices over {%s} the interval comes to ",
      "include no effect: the direction holds throughout, but whether the effect is resolved depends on ",
      "the pooling commitments"), paste(axis_words, collapse = ", ")),
    unstable = sprintf(paste0("the exposure estimate changes sign under some pool-or-stratify choice over ",
      "{%s}; the conclusion depends on these pooling commitments and does not hold across the lattice"),
      paste(axis_words, collapse = ", ")))

  audit$lattice <- list(axes = axes, variables = as.list(axis_var), nodes = nodes,
                        verdict = verdict, reading = reading)
  audit
}
